fuzzers/main.py

```diff
@@ -0,0 +1,819 @@
+        Signal the fuzzer process (with all its process group as a fuzzer can spawn sub-fuzzers)
+        """
+
+        print(f'Sending signal {sig} to fuzzer {self.proc.pid}')
+        os.killpg(os.getpgid(self.proc.pid), sig)
+
+
+# Actually this should also require a ['weight']: int bound but it seems like a mess to encode
+T = typing.TypeVar('T')  # pylint: disable=invalid-name
+
+
+def random_weighted(array: typing.List[T], num: int) -> typing.List[T]:
+    untyped_array = typing.cast(typing.Any, array)
+    return random.choices(array, [x['weight'] for x in untyped_array], k=num)
+
+
+def pause_exit_spot(pause_evt: threading.Event, resume_evt: threading.Event,
+                    exit_evt: threading.Event) -> bool:
+    """
+    Poll the events, returning True if `exit_evt` was set and pausing as requested by `pause_evt`
+    and `resume_evt`
+    """
+
+    resume_evt.clear()
+    if pause_evt.is_set():
+        while not resume_evt.wait(timeout=1):
+            if exit_evt.is_set():
+                return True
+        pause_evt.clear()
+    return exit_evt.is_set()
+
+
+def kill_fuzzers(bucket: gcs.Bucket,
+                 fuzzers: typing.Iterable[FuzzProcess]) -> None:
+    for fuzzer in fuzzers:
+        try:
+            fuzzer.signal(signal.SIGTERM)
+        except ProcessLookupError:
+            print(f'Failed looking up process {fuzzer.proc.pid}')
+    time.sleep(5)
+    for fuzzer in fuzzers:
+        try:
+            fuzzer.signal(signal.SIGKILL)
+        except ProcessLookupError:
+            print(f'Failed looking up process {fuzzer.proc.pid}')
+    for fuzzer in fuzzers:
+        bucket.blob(f'logs/{fuzzer.log_relpath}').upload_from_filename(
+            str(LOGS_DIR / fuzzer.log_relpath))
+
+
+def run_fuzzers(gcs_client: gcs.Client, pause_evt: threading.Event,
+                resume_evt: threading.Event, exit_evt: threading.Event) -> None:
+    """
+    Run all the fuzzers until `exit_evt` gets triggered, pausing and resuming
+    them based on `pause_evt` and `resume_evt`.
+    """
+    # pylint: disable=too-many-locals,too-many-branches,too-many-statements
+
+    bucket = gcs_client.bucket(GCS_BUCKET)
+
+    repo = Repository(REPO_DIR, REPO_URL)
+    repo.clone_if_need_be()
+
+    corpus = Corpus(CORPUS_DIR, bucket)
+
+    while True:
+        if pause_exit_spot(pause_evt, resume_evt, exit_evt):
+            return
+
+        sync_log_files = []
+        date = datetime.datetime.now().strftime('%Y-%m-%d')
+
+        # Read the configuration from the repository
+        cfg = repo.latest_config('master')
+        cfg_for = {
+            b['name']: repo.latest_config(b['name']) for b in cfg['branch']
+        }
+
+        # Figure out which targets we want to run
+        branches: typing.List[BranchType] = random_weighted(
+            cfg['branch'], NUM_FUZZERS)
+        targets: typing.List[TargetType] = [
+            random_weighted(cfg_for[b['name']]['target'], 1)[0]
+            for b in branches
+        ]
```

Comment:
You shouldn’t need to explicitly type the variables here?

Reply:
Right, that's a leftover from the time when I was struggling to understand a typing error, until I figured out the type eg. for `t` in `for t in foo` escapes the scope of the `for` due to python's scoping rules and thus can't be used as another type in another `for`.

TLDR: removed :)